woocommerce/woocommerce
# Add cache refreshing/debugging mechanisms to Report API endpoints

### All Submissions:


### Changes proposed in this Pull Request:

Adds a new collection parameter to all Reports API endpoints that utilize caching, `force_cache_refresh`, which will cause the current request to bypass the cache, re-run the queries for the requested data, and overwrite the previous cache entry with the new results.

Note that this doesn't invalidate the entire cache, only the entry for the particular set of collection parameters and values specified in the request.

This also adds a way to include debugging information related to the cache in the API response. Modeled after the way the Query Monitor plugin [adds such information](https://querymonitor.com/blog/2021/05/debugging-wordpress-rest-api-requests/), you can get this by including an `_envelope` parameter in your API request. The debugging info includes whether the cache has been disabled via filter (`should_use_cache`), whether the `force_cache_refresh` parameter was used, whether the returned data was a `cache_hit` or not, and an array of the query parameters that were actually used to create the cache key.

Closes #33221 

### How to test the changes in this Pull Request:

1. Using an API client like Insomnia, choose one of the WC Analytics API endpoints that this PR updates, and send a request to it. It's best to include the `before` and `after` parameters when making these requests so that you get a consistent time range. Otherwise it defaults to the current second, and you'll never get a cache hit.
2. Now make the request again, but include the `_envelope` parameter. The response should be an object with `body`, `status`, and `headers` properties, along with a new `debug_cache` property.
3. Make this same request a couple of times so you can see that `debug_cache.cache_hit` is set to `true`.
4. Now add the `force_cache_refresh` parameter to the request, with its value set to `true`. With this request, `debug_cache` should show that `cache_hit` is now false.

### Other information:




### FOR PR REVIEWER ONLY:
